### ISSUE TITLE
for `chembl.drug_mechanisms`, can we add CHEMBL.TARGET ID mapping to UniProt?

### ISSUE THREAD

**Issue by colleenXu** [biothings/mychem.info#151]
I asked here https://github.com/biothings/pending.api/issues/100 whether we could make a new pending API for the chembl drug mechanisms data https://mychem.info/v1/query?q=_exists_:%22chembl.drug_mechanisms%22. 

One reason was because the current API used CHEMBL.TARGET IDs (this target entity is involved in the drug mechanism of the chembl compound X). It looks like the CHEMBL.TARGETs are mostly Gene/Protein entities. It would be easier to use if we used a more-universal ID namespace...

Related: Chunlei has found mappings between CHEMBL.TARGET and UniProtKB IDs here https://github.com/biothings/mygene.info/issues/105#issuecomment-1362231645

**Comment by colleenXu** [comment]
Example from Andrew:

The chemical Evacetrapib has a [chembl drug mechanism entry](https://mychem.info/v1/chem/IHIUGIVXARLYHP-YBXDKENTSA-N?fields=chembl,drugcentral) that says the target is `CHEMBL3572` aka `Cholesteryl ester transfer protein` (SINGLE PROTEIN in Homo sapiens). 

The Chembl website shows a [UniProt accession](https://www.ebi.ac.uk/chembl/g/#search_results/all/query=CHEMBL3572) for this CHEMBL3572: [P11597](http://www.uniprot.org/uniprot/?query=accession:P11597). So it 
![Screen Shot 2022-12-28 at 1 26 49 PM](https://user-images.githubusercontent.com/43731687/209873927-5d0ad1c2-2158-444a-b514-6519e5970ba0.png)
seems like CHEMBL.TARGET -> UniProtKB mappings exist.

**Comment by erikyao** [comment]
Related code: https://github.com/biothings/mychem.info/blob/master/src/hub/dataload/sources/chembl/chembl_parser.py#L688

**Comment by erikyao** [comment]
## Source Data

It's not necessary to use the csv/tsv that Chunlei suggested. We can collect the `target["target_components"]["accession"]` fields for each `target` parsed from the already dumped https://www.ebi.ac.uk/chembl/api/data/target.json files.

E.g. in the dumped `target.json` file, `CHEMBL1995` has only one component, which has the accession ID `P43119`:

```python
{
    'organism': 'Homo sapiens',
    'pref_name': 'Prostanoid IP receptor',
    'target_chembl_id': 'CHEMBL1995',
    'target_components': [
        {
            'accession': 'P43119',
            'component_description': 'Prostacyclin receptor',
            'component_id': 325,
            'component_type': 'PROTEIN',
            'relationship': 'SINGLE PROTEIN',
        }
    ],
    'target_type': 'SINGLE PROTEIN',
    'tax_id': 9606
}
```

The result is identical to the accession shown on https://www.ebi.ac.uk/chembl/g/#search_results/all/query=CHEMBL1995

## Special Cases

1. Each target Chembl can have zero, one, or multiple component accessions.
2. `None` accession ID exists. 
    - E.g. [CHEMBL2364096](https://www.ebi.ac.uk/chembl/g/#search_results/all/query=CHEMBL2364096)
    - Solution: Discard those values
3. Not all accession found in the `target.json` are UniProt. 
    - E.g. [CHEMBL1615321](https://www.ebi.ac.uk/chembl/g/#search_results/all/query=CHEMBL1615321) has accession [ENSG00000207827](http://www.uniprot.org/uniprot/?query=accession:ENSG00000207827)
    - Solution: Use regex; see [UniProt Help: Accession](https://www.uniprot.org/help/accession_numbers)

## The Fix

The initial fix is committed to branch [issue-151-fix](https://github.com/biothings/mychem.info/compare/master...issue-151-fix)

Sample document (`CHEMBL238804` with target `CHEMBL1995` and component `P43119`):

```python
{
    '_id': 'QXWZQTURMXZVHJ-UHFFFAOYSA-N',
    'chembl': {
        'molecule_chembl_id': 'CHEMBL238804',
        'inchi_key': 'QXWZQTURMXZVHJ-UHFFFAOYSA-N',
        'smiles': 'CC(C)N(CCCCOCC(=O)NS(C)(=O)=O)c1cnc(-c2ccccc2)c(-c2ccccc2)n1',
        'inchi': 'InChI=1S/C26H32N4O4S/c1-20(2)30(16-10-11-17-34-19-24(31)29-35(3,32)33)23-18
-27-25(21-12-6-4-7-13-21)26(28-23)22-14-8-5-9-15-22/h4-9,12-15,18,20H,10-11,16-17,19H2,1-3H3,
(H,29,31)',
        'drug_indications': [ ... ],
        'drug_mechanisms': [
            {
                'action_type': 'AGONIST',
                'mechanism_refs': [
                    {
                        'id': 'label/2015/207947s000lbl.pdf',
                        'type': 'FDA',
                        'url': 
'http://www.accessdata.fda.gov/drugsatfda_docs/label/2015/207947s000lbl.pdf',
                        'FDA': 'label/2015/207947s000lbl.pdf'
                    }
                ],
                'target_chembl_id': 'CHEMBL1995',
                'target_components': ['P43119'],
                'target_type': 'SINGLE PROTEIN',
                'target_organism': 'Homo sapiens',
                'target_name': 'Prostanoid IP receptor'
            }
        ]
    }
}
```

## TODO: field names and structure

Several questions to @colleenXu :

1. Is `target_components` a good field name? Would you prefer, say `target_component_accessions`, or something else?
2. Would you prefer a list of raw accession IDs like `'target_components': ['P43119']`, or something like `{'target_components': ['uniprot': 'P43119']}`, or any other structure?

**Comment by colleenXu** [comment]
@erikyao

Some questions:
* Are the target components always protein/gene IDs? I noticed [here](https://github.com/biothings/pending.api/issues/100#issuecomment-1366950245) that some targets might not be...
    * I don't know if that affects how target field name and structure
    * I think it's okay to keep ensembl gene IDs, but it'll probably be helpful to have a separate field for them like
```
target_components: {
    ensembl_gene: [ 'P43119', ...],
    uniprot_protein: ['ENSG00000207827', ...],
    ...   
}
```
* I also notice that [some drug mechanisms entries don't seem to have info like target name/ID](https://github.com/biothings/pending.api/issues/100#issuecomment-1367068414). Does your new code catch and handle these? 

@andrewsu and @newgene may also have thoughts?

**Comment by colleenXu** [comment]
@erikyao Feedback on the ENSEMBL ID part: 

The ENSG ID and that entity (microRNA 30a) doesn't seem to exist in Uniprot. Which I guess makes sense since [it doesn't seem to code for a protein](https://www.ncbi.nlm.nih.gov/gene/?term=ENSG00000207827)...

I guess the structure I proposed above (which I edited since I made a mistake) would still work...

**Comment by erikyao** [comment]
Thank you, @colleenXu 

> Are the target components always protein/gene IDs?

From the data I parsed from https://www.ebi.ac.uk/chembl/api/data/target.json, a component is either None, an Ensembl Gene ID, or a UniProt accession ID. 

> I guess the structure I proposed above (which I edited since I made a mistake) would still work...

Sure, I can make it in the parser.

**Comment by erikyao** [comment]
> I also notice that https://github.com/biothings/pending.api/issues/100#issuecomment-1367068414. Does your new code catch and handle these?

Nope, my current code cannot fix this problem. Do you have any idea (like a second file/API to fill the blanks)?

**Comment by colleenXu** [comment]
@erikyao 

* Do you think the uniprot IDs are always proteins? If not, you can change the field name I proposed from `uniprot_protein` to `uniprot`...
* Oops I didn't put the right [link](https://mychem.info/v1/query?q=_exists_:chembl.drug_mechanisms%20AND%20(NOT%20_exists_:%22chembl.drug_mechanisms.action_type%22)%20AND%20(NOT%20_exists_:%22chembl.drug_mechanisms.target_chembl_id%22)&fields=chembl) for viewing the 369 chemicals where the drug_mechanisms.action_type is null (and it looks like the target_chembl_id are also null) in my [post](https://github.com/biothings/mychem.info/issues/151#issuecomment-1414433124) so I fixed that. I'm wondering whether these entries make sense to remove entirely, or whether they can be kept and the fields with null values can be removed...

**Comment by erikyao** [comment]
> Do you think the uniprot IDs are always proteins?

I am not sure... and I think the `uniprot` is a better prefix.

> I'm wondering whether these entries make sense to remove entirely, or whether they can be kept and the fields with null values can be removed

My latest code will remove those null field. Those documents will be kept.

**Comment by colleenXu** [comment]
Here's the chembl sites for [some chemicals that have this "null" issue](https://mychem.info/v1/query?q=_exists_:chembl.drug_mechanisms%20AND%20(NOT%20_exists_:%22chembl.drug_mechanisms.action_type%22)%20AND%20(NOT%20_exists_:%22chembl.drug_mechanisms.target_chembl_id%22)&fields=chembl)

Looks like CHEMBL's website also lacks the target info / any mechanism info.
* [CHEMBL3545306](https://www.ebi.ac.uk/chembl/compound_report_card/CHEMBL3545306/)
* https://www.ebi.ac.uk/chembl/compound_report_card/CHEMBL1200346/

**Comment by erikyao** [comment]
> Looks like CHEMBL's website also lacks the target info / any mechanism info.

Yep, I think we can at first align our documents to the the CHEMBL report cards, and then fill the blanks in the future if necessary.

**Comment by erikyao** [comment]
Side note: I found a possible error in CHEMBL, and have reported it to https://github.com/chembl/GLaDOS/issues/1310

**Comment by erikyao** [comment]
Check the latest release of MyChem for `CHEMBL238804`: https://mychem.info/v1/chem/CHEMBL238804?fields=chembl

The `target_components` field has been added.

**Comment by colleenXu** [comment]
Looks good to me. @erikyao, can we close this issue?

**Comment by erikyao** [comment]
@colleenXu issue closed.

One more thing: although we have `target_components` source data containing `ensembl_gene` entries, none of them is mapped to any of our `target` records. Therefore currently we only have `target_components` with `uniprot` in our document collection.

**Comment by colleenXu** [comment]
@erikyao I'm confused. I see 60 documents that seem to have both `target_components.ensembl fields` and `target_chembl_id` fields? https://mychem.info/v1/query?q=_exists_:%22chembl.drug_mechanisms.target_components.ensembl_gene%22%20AND%20_exists_:%22chembl.drug_mechanisms.target_chembl_id%22&fields=chembl

**Comment by erikyao** [comment]
@colleenXu sorry, my bad. I found I was analyzing on a subset of source data... Gladly we have both the uniprot and ensembl fields indexed.
